### ISSUE TITLE
systemd backend: RuntimeError("can't start new thread")

### ISSUE THREAD

**Issue by Xenopathic** [fail2ban/fail2ban#969]
I have noticed a bug when running Fail2ban v0.9.1 on a Raspberry Pi: I can't enable more than 2 jails before none of them start, printing an error in the logs:

```
WARNING Command ['start', 'foobar'] has failed. Received RuntimeError("can't start new thread",)
```

I've tracked it down to the point at which a thread is spawned for the systemd filter. If I remove all references to `self.__journal` in `filtersystemd.py`, the jails will successfully start (albeit not doing anything as the logic is ripped out), so it's definitely not a resource issue. What's more, simply restoring the initialization of `self.__journal` in `__init__`, the error occurs again. I find it odd that it will work fine with 2 systemd jails enabled, but 3 or more and all of the jails fail.

My theory is, that there is either some kind of connection limit enforced by journald that closes all connections if a process exceeds that limit. I'm not an expert on how Python's threading library works, so I don't know how it handles the creation of a new thread with respect to copying variables, but it could be that the start of a new thread triggers a copy of the `systemd.journal.Reader` object, which exceeds the resource limit and the thread creation fails.

This issue is fairly important to me, so I'd be glad to help with testing (and small patches) wherever I can.

**Comment by Xenopathic** [comment]
I have a reproducible test:

```
from systemd import journal
from threading import Thread

class MyThread(Thread):
	def __init__(self):
		super(MyThread, self).__init__()
		self.journal = journal.Reader()

thread1 = MyThread()
thread2 = MyThread()
thread3 = MyThread()
thread1.start()
thread2.start()
thread3.start()
```

On my x86_64 system running Arch Linux and Python 3.4.2, that script executes without any errors. Run it on my Pi (armv6h, Arch with Python 3.4.2) and the same `can't start new thread` error shows up. As with the error in Fail2ban, removing the creation of `thread3` lets the other two threads work correctly. Likewise, removing the creation of the `journal.Reader` makes everything work also.

This is clearly a bug in either Python's threading code or the systemd journal Python module, but I think it'd be best to keep this issue open in case anyone else discovers the same problem.

**Comment by Xenopathic** [comment]
I did some debugging: looks like `pthread_create()` in Python's threading library returns `EAGAIN`, Resource temporarily unavailable. It could mean resource limits are being hit...

**Comment by kakra** [comment]
This is one year old now, did you find a fix? I see it happening as soon as I activate more than one jail.

**Comment by sebres** [comment]
#1513 seems to be the same

**Comment by sebres** [comment]
Can someone having such errors please test following command (to basically check whether the same error occurs), pure despair but who knows:
```bash
PYTHONPATH=. python -c 'import sys, logging, time; logging.basicConfig(stream=sys.stdout, level=logging.DEBUG); from fail2ban.server.filtersystemd import FilterSystemd; filter = FilterSystemd(None); filter2 = FilterSystemd(None); print("** start 2 threads ..."); filter.start(); filter2.start(); print("** wait a little bit ..."); time.sleep(3.0); print("** stop ..."); filter.stop(); filter2.stop(); time.sleep(1.0)'
```

**Comment by sebres** [comment]
I cannot reproduce described behavior with 3 jails using systemd backend neither in python2.7 nor in python3.5... Tested latest 0.9 and 0.10 but also oldest 0.9.3 on debian (jessie) and ubuntu 16.04...

#### So the temporary conclusion: it would be platform, python version or python-systemd module related bug...

Here are my versions where it successful runs:
- fail2ban:&ensp;0.9.3,&ensp;0.9.6,&ensp;0.10.0a2
- python:&ensp;2.7.11-1
- python-systemd:&ensp;231-2build1
- python3:&ensp;3.5.1-3
- python3-systemd:&ensp;231-2build1


Test jail.local:
```
[sshd]
backend = systemd
enabled = true

[pam-generic]
backend = systemd
enabled = true

[dovecot]
journalmatch = _SYSTEMD_UNIT=dovecot.service
backend = systemd
enabled = true
```

A log-file excerpt of successful start (and even some test failures found):
```
2016-08-17 11:40:03,811 fail2ban.server         [19508]: INFO    Starting Fail2ban v0.10.0a2
2016-08-17 11:40:03,812 fail2ban.server         [19508]: INFO    Daemon started
2016-08-17 11:40:03,894 fail2ban.database       [19508]: INFO    Connected to fail2ban persistent database '/var/lib/fail2ban/fail2ban.sqlite3'
2016-08-17 11:40:03,896 fail2ban.jail           [19508]: INFO    Creating new jail 'sshd'
2016-08-17 11:40:03,911 fail2ban.jail           [19508]: INFO    Jail 'sshd' uses systemd
2016-08-17 11:40:03,913 fail2ban.jail           [19508]: INFO    Initiated 'systemd' backend
2016-08-17 11:40:03,915 fail2ban.filter         [19508]: INFO    Set maxRetry = 5
2016-08-17 11:40:03,917 fail2ban.actions        [19508]: INFO    Set banTime = 600
2016-08-17 11:40:03,918 fail2ban.filter         [19508]: INFO    Set findtime = 600
2016-08-17 11:40:03,918 fail2ban.filter         [19508]: INFO    Set maxlines = 10
2016-08-17 11:40:03,985 fail2ban.filtersystemd  [19508]: INFO    Added journal match for: '_SYSTEMD_UNIT=sshd.service + _COMM=sshd'
2016-08-17 11:40:03,988 fail2ban.jail           [19508]: INFO    Creating new jail 'dovecot'
2016-08-17 11:40:03,988 fail2ban.jail           [19508]: INFO    Jail 'dovecot' uses systemd
2016-08-17 11:40:03,988 fail2ban.jail           [19508]: INFO    Initiated 'systemd' backend
2016-08-17 11:40:03,989 fail2ban.filter         [19508]: INFO    Set maxRetry = 5
2016-08-17 11:40:03,991 fail2ban.actions        [19508]: INFO    Set banTime = 600
2016-08-17 11:40:03,992 fail2ban.filter         [19508]: INFO    Set findtime = 600
2016-08-17 11:40:04,017 fail2ban.filtersystemd  [19508]: INFO    Added journal match for: '_SYSTEMD_UNIT=dovecot.service'
2016-08-17 11:40:04,019 fail2ban.jail           [19508]: INFO    Creating new jail 'pam-generic'
2016-08-17 11:40:04,019 fail2ban.jail           [19508]: INFO    Jail 'pam-generic' uses systemd
2016-08-17 11:40:04,020 fail2ban.jail           [19508]: INFO    Initiated 'systemd' backend
2016-08-17 11:40:04,021 fail2ban.filter         [19508]: INFO    Set maxRetry = 5
2016-08-17 11:40:04,023 fail2ban.actions        [19508]: INFO    Set banTime = 600
2016-08-17 11:40:04,023 fail2ban.filter         [19508]: INFO    Set findtime = 600
2016-08-17 11:40:04,033 fail2ban.filter         [19508]: INFO    [sshd] Found 172.18.105.48 - 2016-08-17 11:31:46
2016-08-17 11:40:04,034 fail2ban.jail           [19508]: INFO    Jail 'sshd' started
2016-08-17 11:40:04,036 fail2ban.filter         [19508]: INFO    [sshd] Found 172.18.105.48 - 2016-08-17 11:31:49
2016-08-17 11:40:04,037 fail2ban.filter         [19508]: INFO    [sshd] Found 172.18.105.48 - 2016-08-17 11:31:51
2016-08-17 11:40:04,042 fail2ban.jail           [19508]: INFO    Jail 'dovecot' started
2016-08-17 11:40:04,044 fail2ban.filtersystemd  [19508]: NOTICE  Jail started without 'journalmatch' set. Jail regexs will be checked against all journal entries, which is not advised for performance reasons.
2016-08-17 11:40:04,045 fail2ban.jail           [19508]: INFO    Jail 'pam-generic' started
2016-08-17 11:40:04,047 fail2ban.filter         [19508]: INFO    [pam-generic] Found 172.18.105.48 - 2016-08-17 11:31:49
...
2016-08-17 11:40:10,398 fail2ban.server         [19508]: INFO    Stopping all jails
2016-08-17 11:40:10,474 fail2ban.jail           [19508]: INFO    Jail 'dovecot' stopped
2016-08-17 11:40:10,533 fail2ban.jail           [19508]: INFO    Jail 'sshd' stopped
2016-08-17 11:40:10,663 fail2ban.jail           [19508]: INFO    Jail 'pam-generic' stopped
2016-08-17 11:40:10,664 fail2ban.server         [19508]: INFO    Exiting Fail2ban
```

**Comment by knappmk** [comment]
My Rasperry Pi 2 ran into the same situation as @Xenopathic descripted. My OS is "Arch Linux". I already tried serveral things. I installed "fail2ban" from the package manager (0.9.5-1) and tried version 0.10.0a1. There's no change in behavoir, when I downgrade from "python 3.5.2-1" to "python 3.5.1-2" and when I install "fail2ban" with "python2 2.7.12-1". All dependencies listed on the package page on "archlinux.com" were installed (for python2 and python3 [python(2)-systemd 231-1]).

Running the testing code of @Xenopathic doesn't generate any errors.

Executing the code of @sebres ...
<pre># python3 -c 'import sys, logging, time; logging.basicConfig(stream=sys.stdout, level=logging.DEBUG); from fail2ban.server.filtersystemd import FilterSystemd; filter = FilterSystemd(None); filter2 = FilterSystemd(None); print("** start 2 threads ..."); filter.start(); filter2.start(); print("** wait a little bit ..."); time.sleep(3.0); print("** stop ..."); filter.stop(); filter2.stop(); time.sleep(1.0)'
DEBUG:fail2ban.filter:Setting usedns = warn for FilterSystemd(None)
DEBUG:fail2ban.filter:Created FilterSystemd(None)
DEBUG:fail2ban.filtersystemd:Created FilterSystemd
DEBUG:fail2ban.filter:Setting usedns = warn for FilterSystemd(None)
DEBUG:fail2ban.filter:Created FilterSystemd(None)
DEBUG:fail2ban.filtersystemd:Created FilterSystemd
** start 2 threads ...
NOTICE:fail2ban.filtersystemd:Jail started without 'journalmatch' set. Jail regexs will be checked against all journal entries, which is not advised for performance reasons.
NOTICE:fail2ban.filtersystemd:Jail started without 'journalmatch' set. Jail regexs will be checked against all journal entries, which is not advised for performance reasons.
** wait a little bit ...
** stop ...
DEBUG:fail2ban.filtersystemd:jailless filter terminated
DEBUG:fail2ban.filtersystemd:jailless filter terminated</pre>
... dosen't change anything.

The errors also occour using this setup:

- python2 2.7.11-1

- python2-systemd 231-1

- fail2ban 0.9.3

<pre>[...]
fail2ban.transmitter[5022]: WARNING Command ['start', 'sshd'] has failed. Received error("can't start new thread",)
fail2ban-client[4976]: ERROR  NOK: ("can't start new thread",)
fail2ban.transmitter[5022]: WARNING Command ['start', 'sshd-ddos'] has failed. Received error("can't start new thread",)
fail2ban-client[4976]: ERROR  NOK: ("can't start new thread",)
[...]</pre>

On my system the errors occur when I enable 4 or more 'systemd jails', jails using log-files don't influence the problem (well already guesst from posts before?).

**Comment by sebres** [comment]
> Executing the code of @sebres ... ... doesn't change anything.

That should at all not change something - it is just a test script, that creates exactly 2 thread instances used systemd backend, to test it can work at all (because original issue says about 2 systemd jails);

As I understood you, in your case it is 4 or more threads...

**Comment by blackPantherOS** [comment]
I have same problem with fail2ban 0.9.6 if I enable six or more jail : 

ERROR  NOK: ("can't start new thread",)
ERROR  NOK: ("can't start new thread",)
ERROR  NOK: ("can't start new thread",)
ERROR  NOK: ("can't start new thread",)
ERROR  NOK: ("can't start new thread",)
ERROR  NOK: ("can't start new thread",)
ERROR  NOK: ("can't start new thread",)

python3-3.4.1 and 3.5.1 as well

**Comment by Farbdose** [comment]
One more PI with this problem:

- OS: Arch Linux
- fail2ban: 0.9.6
- python3: 3.6.0
- python3-systemd: 231

@sebres I had to increase the thread count to 3 to trigger the error and ended up making it editable:

```python
python -c 'threads=3; import sys, logging, time; logging.basicConfig(stream=sys.stdout, level=logging.DEBUG); from fail2ban.server.filtersystemd import FilterSystemd; filters = [FilterSystemd(None) for i in range(threads)]; print("** start "+str(len(filters))+" threads ..."); [f.start() for f in filters]; print("** wait a little bit ..."); time.sleep(3.0); print("** stop ..."); [f.stop() for f in filters[::-1]]; time.sleep(1.0)'
```

Executing the test script now results in:
```python
DEBUG:fail2ban.filter:Setting usedns = warn for FilterSystemd(None)
DEBUG:fail2ban.filter:Created FilterSystemd(None)
DEBUG:fail2ban.filtersystemd:Created FilterSystemd
DEBUG:fail2ban.filter:Setting usedns = warn for FilterSystemd(None)
DEBUG:fail2ban.filter:Created FilterSystemd(None)
DEBUG:fail2ban.filtersystemd:Created FilterSystemd
DEBUG:fail2ban.filter:Setting usedns = warn for FilterSystemd(None)
DEBUG:fail2ban.filter:Created FilterSystemd(None)
DEBUG:fail2ban.filtersystemd:Created FilterSystemd
** start 3 threads ...
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "<string>", line 1, in <listcomp>
  File "/usr/lib/python3.6/site-packages/fail2ban/server/jailthread.py", line 81, in start
    super(JailThread, self).start()
  File "/usr/lib/python3.6/threading.py", line 846, in start
    _start_new_thread(self._bootstrap, ())
RuntimeError: can't start new thread
```

Please provide further debugging steps so I can help you track it down.

**Comment by sebres** [comment]
> something is eating the thread stack 

I don't think so, but I'm neither an author of python systemd module (nor corresponding libraries), therefore I'm not really sure...

> or the initial stack size is awfully low

Rather this is true

Which default values of the stack-limits (soft/hard) did you have in your RasPI?

**Comment by Farbdose** [comment]
stack size limits are:

- soft: 8192 kbytes
- hard: unlimited

Looks pretty normal to me.

**Comment by sebres** [comment]
> Is there a python config for these limits other than the system defaults?

No idea...

**Comment by Xenopathic** [comment]
@Farbdose I had to up the number of threads to 4 to get the RuntimeError, but your stack size hack worked to fix it. My stack sizes as reported by `ulimit` are the same as you too.

**Comment by Farbdose** [comment]
@sebres 
>Is there a python config for these limits other than the system defaults?

After some searching I came to conclusion: no there isn't.
So we are back to square one.

**Comment by sebres** [comment]
> something is eating the thread stack

> stack_size(32*1024)

I think you'll rather reduce the stack size used per thread, if you do `stack_size(32*1024)` (it's a 32KB), so you can then theoretically create 256 threads.

The question is: Why your python have too large initial value of thread stack size or perhaps why it uses another (larger?) granularity on Raspbian as on the "normal" debian?

**Comment by jPPuetz** [comment]
Just wanted to notice that this issue is still present and the workaround from @Farbdose still works.
- RPI 3 B
- Arch Linux x86
- Fail2Ban v0.10.4
- Python 3.7.2
- python-systemd-234-2

**Comment by jPPuetz** [comment]
Just wanted to notice that this issue is still present and the workaround from @Farbdose does not work for me. After adding the two lines to jailthread.py fail2ban restarts every 2 seconds in an inifinite loop.
- RPI 3 B
- Arch Linux x86
- Fail2Ban v0.10.4
- Python 3.7.2
- python-systemd-234-2

**Comment by BrozzSama** [comment]
Yes I have the same exact behavior as @jPPuetz using an odroid-c1 and arch linux arm.

**Comment by sebres** [comment]
merged, see #2356
